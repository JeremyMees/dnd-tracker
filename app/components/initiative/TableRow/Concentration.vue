<script setup lang="ts">
const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}>()

function updateConcentration(): void {
  if (!props.sheet) return
  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = [...props.sheet.rows]

  if (index === -1 || !rows[index]) return

  rows[index] = { ...rows[index], concentration: !props.item.concentration }

  props.update({ rows })
}
</script>

<template>
  <button
    v-if="item.type !== 'lair'"
    v-tippy="$t('general.concentration')"
    :aria-label="$t('general.concentration')"
    class="icon-btn-primary mx-auto"
    @click="updateConcentration"
  >
    <Icon
      :name="`tabler:${item.concentration ? 'circle-filled' : 'circle-dotted'}`"
      :aria-hidden="true"
      class="size-5 min-w-5"
    />
  </button>
</template>
