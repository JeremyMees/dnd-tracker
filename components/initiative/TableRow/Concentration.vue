<script setup lang="ts">
const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
}>()

function updateConcentration(): void {
  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = props.sheet.rows

  if (index === -1) return

  rows[index].concentration = !rows[index].concentration

  props.update({ rows })
}
</script>

<template>
  <td>
    <button
      v-if="item.type !== 'lair'"
      v-tippy="$t('general.concentration')"
      :aria-label="$t('general.concentration')"
      class="icon-btn-primary mx-auto"
      @click="updateConcentration"
    >
      <Icon
        :name="`tabler:${item.concentration ? 'brain' : 'point'}`"
        :aria-hidden="true"
        class="size-5 min-w-5"
      />
    </button>
  </td>
</template>
