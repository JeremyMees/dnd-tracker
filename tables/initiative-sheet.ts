import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
