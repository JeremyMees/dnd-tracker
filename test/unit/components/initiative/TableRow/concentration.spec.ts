import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import Concentration from '~/components/initiative/TableRow/Concentration.vue'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'

interface Props {
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}

const mockUpdate = vi.fn()

const props: Props = {
  item: sheet.rows[0]!,
  sheet,
  update: mockUpdate,
}

describe('Initiative table row concentration', async () => {
  beforeEach(() => mockUpdate.mockClear())

  it('Should match snapshot', async () => {
    const component = await mountSuspended(Concentration, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should not render button for lair type', async () => {
    const component = await mountSuspended(Concentration, {
      props: {
        ...props,
        item: { ...props.item, type: 'lair' as const },
      },
    })

    expect(component.find('button').exists()).toBeFalsy()
  })

  it('Should render button for non-lair type', async () => {
    const component = await mountSuspended(Concentration, { props })

    expect(component.find('button').exists()).toBeTruthy()
  })

  it('Should show filled circle when concentration is true', async () => {
    const component = await mountSuspended(Concentration, { props })

    expect(component.find('[data-test-concentration="true"]').exists()).toBeTruthy()
  })

  it('Should show dotted circle when concentration is false', async () => {
    const component = await mountSuspended(Concentration, {
      props: {
        ...props,
        item: { ...props.item, concentration: false },
      },
    })

    expect(component.find('[data-test-concentration="false"]').exists()).toBeTruthy()
  })

  it('Should call update with toggled concentration', async () => {
    const component = await mountSuspended(Concentration, { props })

    await component.find('button').trigger('click')

    expect(mockUpdate).toHaveBeenCalledWith({
      rows: expect.arrayContaining([
        expect.objectContaining({
          concentration: !props.item.concentration,
        }),
      ]),
    })
  })

  it('Should not call update when sheet is undefined', async () => {
    const component = await mountSuspended(Concentration, {
      props: {
        ...props,
        sheet: undefined,
      },
    })

    await component.find('button').trigger('click')

    expect(mockUpdate).not.toHaveBeenCalled()
  })
})
