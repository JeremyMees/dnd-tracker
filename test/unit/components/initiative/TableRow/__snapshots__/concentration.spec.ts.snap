// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative table row concentration > Should match snapshot 1`] = `"<button data-test-concentration="true" aria-label="general.concentration" class="icon-btn-primary mx-auto"><span class="iconify i-tabler:circle-filled size-5 min-w-5" aria-hidden="true"></span></button>"`;
